### ISSUE TITLE
cannot set OUTTYPE of some user-defined variables; TMAP ERR on resetting OUTTYPE

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#918]
**Reported by andrew wittenberg on 5 Mar 2009 20:28 UTC**
```
        NOAA/PMEL TMAP
        FERRET v6.192
        Linux(g77) 2.4.21-32 - 02/13/09
         5-Mar-09 15:01

! Why can't we set the outtype of a user-defined variable?

yes? let a = {1,2,3}
yes? set var/outtype=double a
 *** NOTE: Converting single precision variable to double on output
 **ERROR: command syntax:
          illegal character:
yes? save/clob/file=a.nc a
 LISTing to file a.nc
yes? sp ncdump -h a.nc
netcdf a {
dimensions:
        XAX1_3 = 3 ;
variables:
        double XAX1_3(XAX1_3) ;
                XAX1_3:point_spacing = "even" ;
                XAX1_3:axis = "X" ;
        float A(XAX1_3) ;
                A:missing_value = -1.e+34f ;
                A:_FillValue = -1.e+34f ;
                A:long_name = "{1,2,3}" ;

// global attributes:
                :history = "FERRET V6.192    5-Mar-09" ;
                :Conventions = "CF-1.0" ;
}


! But we can correctly write a file-variable to DOUBLE:

yes? can var a; use a.nc
yes? set var/outtype=double a
 *** NOTE: Converting single precision variable to double on output
yes? save/clob/file=b.nc a
 LISTing to file b.nc
 *** NOTE: Converting data type of missing_value NC_FLOAT to match output type
of variable NC_DOUBLE
yes? sp ncdump -h b.nc
netcdf b {
dimensions:
        XAX1_3 = 3 ;
variables:
        double XAX1_3(XAX1_3) ;
                XAX1_3:point_spacing = "even" ;
                XAX1_3:axis = "X" ;
        double A(XAX1_3) ;
                A:missing_value = -9.99999979021477e+33 ;
                A:_FillValue = -9.99999979021477e+33 ;
                A:long_name = "{1,2,3}" ;
                A:Ferret_Precision_Note = "This variable written by Ferret was
converted from FLOAT to DOUBLE" ;
                A:history = "From a" ;

// global attributes:
                :history = "FERRET V6.192    5-Mar-09" ;
                :Conventions = "CF-1.0" ;
}

! Where is the following error coming from?

yes? let c = a
yes? set var/outtype=float a
yes? set var/outtype=double a                                                                   
 *** NOTE: Converting single precision variable to double on output                             
 **TMAP ERR: attribute already defined                                                          
             Ferret_Precision_Note for variable A                                               
             Data set: ./a.nc                                                                   

```

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1646

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 16 Mar 2009 19:39 UTC**
This is fixed; the operation is allowed for user-defined variables, but was incorrectly trying to search for the dataset.

```
yes? let a = {1,2,3}
yes? SET VAR/OUTTYPE = double a
 *** NOTE: Converting single precision variable to double on output
yes? SAVE/CLOBBER/FILE=a.nc a
 LISTing to file a.nc
yes? SP ncdump a.nc
netcdf a {
dimensions:
        XAX1_3 = 3 ;
variables:
        double XAX1_3(XAX1_3) ;
                XAX1_3:point_spacing = "even" ;
                XAX1_3:axis = "X" ;
        double A(XAX1_3) ;
                A:missing_value = -9.99999979021477e+33 ;
                A:_FillValue = -9.99999979021477e+33 ;
                A:long_name = "{1,2,3}" ;
                A:Ferret_Precision_Note = "This variable written by Ferret was converted from FLOAT to DOUBLE" ;

// global attributes:
                :history = "FERRET V6.193   16-Mar-09" ;
                :Conventions = "CF-1.0" ;
data:

 XAX1_3 = 1, 2, 3 ;

 A = 1, 2, 3 ;
}
```

Or, the float-to-double conversion looks like this:

```

yes? LET a = {1.1, 2.2, 3.3}

yes? SET VAR/OUTTYPE = double a
 *** NOTE: Converting single precision variable to double on output
yes? SAVE/CLOBBER/FILE=a.nc a
 LISTing to file a.nc
yes? SP ncdump a.nc
netcdf a {
dimensions:
        XAX1_3 = 3 ;
variables:
        double XAX1_3(XAX1_3) ;
                XAX1_3:point_spacing = "even" ;
                XAX1_3:axis = "X" ;
        double A(XAX1_3) ;
                A:missing_value = -9.99999979021477e+33 ;
                A:_FillValue = -9.99999979021477e+33 ;
                A:long_name = "{1.1,2.2,3.3}" ;
                A:Ferret_Precision_Note = "This variable written by Ferret was converted from FLOAT to DOUBLE" ;

// global attributes:
                :history = "FERRET V6.193   16-Mar-09" ;
                :Conventions = "CF-1.0" ;
data:

 XAX1_3 = 1, 2, 3 ;

 A = 1.10000002384186, 2.20000004768372, 3.29999995231628 ;
}
```

Or, convert a variable to integers:

```

yes? LET a = { 3.77, 4.00, 4.10, 4.33, 4.38, 4.27, 4.31, 4.40, 4.75, 5.01, 5.27, 5.52}
** INTERRUPTED! **
yes? SET VAR/BAD=999/OUTTYPE=int a
yes? SAVE/CLOBBER/FILE=a.nc a
 LISTing to file a.nc
yes? SP ncdump a.nc
netcdf a {
dimensions:
        XAX1_12 = 12 ;
variables:
        double XAX1_12(XAX1_12) ;
                XAX1_12:point_spacing = "even" ;
                XAX1_12:axis = "X" ;
        int A(XAX1_12) ;
                A:missing_value = 999 ;
                A:_FillValue = 999 ;
                A:long_name = "{ 3.77, 4.00, 4.10, 4.33, 4.38, 4.27, 4.31, 4.40, 4.75, 5.01, 5.27, 5.52}" ;

// global attributes:
                :history = "FERRET V6.193   16-Mar-09" ;
                :Conventions = "CF-1.0" ;
data:

 XAX1_12 = 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 ;

 A = 3, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5 ;
}



```

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 7 Jul 2009 15:54 UTC**
This is almost fixed as of v6.203, though I still can't change the OUTTYPE back to DOUBLE again, if I change back to FLOAT:

```
        NOAA/PMEL TMAP
        FERRET v6.203  
        Linux(g77) 2.4.21-32 - 06/26/09
         7-Jul-09 11:48     

yes? let a = {1,2,3}
yes? set var/outtype=double a
 *** NOTE: Converting single precision variable to double on output
yes? set var/outtype=float a
yes? set var/outtype=double a
 *** NOTE: Converting single precision variable to double on output
 **TMAP ERR: attribute already defined
             Ferret_Precision_Note for variable a
             Data set: 
yes? save/clob/file=a.nc a
 LISTing to file a.nc
yes? sp ncdump -h a.nc 
netcdf a {
dimensions:
        XAX1_3 = 3 ;
variables:
        double XAX1_3(XAX1_3) ;
                XAX1_3:point_spacing = "even" ;
                XAX1_3:axis = "X" ;
        float A(XAX1_3) ;
                A:missing_value = -1.e+34f ;
                A:_FillValue = -1.e+34f ;
                A:long_name = "{1,2,3}" ;
                A:Ferret_Precision_Note = "This variable written by Ferret was converted from FLOAT to DOUBLE" ;

// global attributes:
                :history = "FERRET V6.203    7-Jul-09" ;
                :Conventions = "CF-1.0" ;
}

```

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 16 Nov 2009 22:27 UTC**
This now appears to be fixed in v6.31beta (dated 11/5/09).
